src/cli/cli_tcp.cpp

```diff
@@ -0,0 +1,522 @@
+    VerifyOrExit(command != nullptr, error = OT_ERROR_INVALID_COMMAND);
+
+    error = (this->*command->mHandler)(aArgs + 1);
+
+exit:
+    return error;
+}
+
+void TcpExample::HandleTcpEstablishedCallback(otTcpEndpoint *aEndpoint)
+{
+    static_cast<TcpExample *>(otTcpEndpointGetContext(aEndpoint))->HandleTcpEstablished(aEndpoint);
+}
+
+void TcpExample::HandleTcpSendDoneCallback(otTcpEndpoint *aEndpoint, otLinkedBuffer *aData)
+{
+    static_cast<TcpExample *>(otTcpEndpointGetContext(aEndpoint))->HandleTcpSendDone(aEndpoint, aData);
+}
+
+void TcpExample::HandleTcpReceiveAvailableCallback(otTcpEndpoint *aEndpoint,
+                                                   size_t         aBytesAvailable,
+                                                   bool           aEndOfStream,
+                                                   size_t         aBytesRemaining)
+{
+    static_cast<TcpExample *>(otTcpEndpointGetContext(aEndpoint))
+        ->HandleTcpReceiveAvailable(aEndpoint, aBytesAvailable, aEndOfStream, aBytesRemaining);
+}
+
+void TcpExample::HandleTcpDisconnectedCallback(otTcpEndpoint *aEndpoint, otTcpDisconnectedReason aReason)
+{
+    static_cast<TcpExample *>(otTcpEndpointGetContext(aEndpoint))->HandleTcpDisconnected(aEndpoint, aReason);
+}
+
+otTcpIncomingConnectionAction TcpExample::HandleTcpAcceptReadyCallback(otTcpListener *   aListener,
+                                                                       const otSockAddr *aPeer,
+                                                                       otTcpEndpoint **  aAcceptInto)
+{
+    return static_cast<TcpExample *>(otTcpListenerGetContext(aListener))
+        ->HandleTcpAcceptReady(aListener, aPeer, aAcceptInto);
+}
+
+void TcpExample::HandleTcpAcceptDoneCallback(otTcpListener *   aListener,
+                                             otTcpEndpoint *   aEndpoint,
+                                             const otSockAddr *aPeer)
+{
+    static_cast<TcpExample *>(otTcpListenerGetContext(aListener))->HandleTcpAcceptDone(aListener, aEndpoint, aPeer);
+}
+
+void TcpExample::HandleTcpEstablished(otTcpEndpoint *aEndpoint)
+{
+    OT_UNUSED_VARIABLE(aEndpoint);
+    mInterpreter.OutputLine("TCP: Connection established");
+}
+
+void TcpExample::HandleTcpSendDone(otTcpEndpoint *aEndpoint, otLinkedBuffer *aData)
+{
+    OT_UNUSED_VARIABLE(aEndpoint);
+
+    if (mBenchmarkBytesTotal == 0)
+    {
+        /*
+         * If the benchmark encountered an error, we might end up here. So,
+         * tolerate some benchmark links finishing in this case.
+         */
+        if (aData == &mSendLink)
+        {
+            OT_ASSERT(mSendBusy);
+            mSendBusy = false;
+        }
+    }
+    else
+    {
+        OT_ASSERT(aData != &mSendLink);
+        mBenchmarkLinksLeft--;
+        if (mBenchmarkLinksLeft >= OT_ARRAY_LENGTH(mBenchmarkLinks))
+        {
+            if (otTcpSendByReference(&mEndpoint, aData, 0) != OT_ERROR_NONE)
```

Comment:
What if the `SendDone` is for the first entry, i.e. `mBenchmarkLinks[0]`which can have a length smaller than other ones?

Should we change the `aData.mLength`?

Reply:
Good point, I will update. I suppose this bug didn't show up in my testing because, `mBenchmarkBytesTotal` was a multiple of `sizeof(mSendBuffer)`, so I will do a bit more testing there as well.